src/Adapters/Auth/index.js

```diff
@@ -56,15 +59,40 @@ function authDataValidator(adapter, appIds, options) {
   }
 }
 
+function validateAdapter(adapter) {
+  const mismatches = Object.getOwnPropertyNames(AuthAdapter.prototype).reduce((obj, key) => {
+    const adapterType = typeof adapter[key];
+    const expectedType = typeof AuthAdapter.prototype[key];
+    if (adapterType !== expectedType) {
+      obj[key] = {
+        expected: expectedType,
+        actual: adapterType
+      }
+    }
+    return obj;
+  }, {});
+
+  let ret = null;
+  if (Object.keys(mismatches).length > 0) {
+    ret = new Error('Adapter prototype doesn\'t match expected prototype', adapter, mismatches);
+  }
+  return ret;
+}
+
 function loadAuthAdapter(provider, authOptions) {
-  const defaultAdapter = providers[provider];
-  const adapter = Object.assign({}, defaultAdapter);
+  let defaultAdapter = providers[provider];
   const providerOptions = authOptions[provider];
+  if (providerOptions && providerOptions.hasOwnProperty('oauth2')) {
```

Comment:
you should also check that the `defaultAdapter` is undefined. What if a 3rd party adapter needs the `oauth2` option set to something, also please merge the if in the next line with this one.

Reply:
> you should also check that the defaultAdapter is undefined

I'm not sure I understand this right.
A few lines after the one you commented on comes this:
```
  if (!defaultAdapter && !providerOptions) {
    return;
  }
```

This already checks that defaultAdapter actually points to something.
So if loadAuthAdapter() is called with an adapter name that is not defined in the "providers" object (as a property), then the "if (!defaultAdapter ...)" check should catch it.

The above condition verification is part of the current Parse code on the master branch. But now that I think about it, it's probably wrong, since it'll skip loading the adapter only if an adapter implementation is not found and no options are specified in the config. Whereas it should only check whether an adapter implementation is available or not.

So the proper check would be:
```
  if (!defaultAdapter) {
    return;
  }
```

On the other hand, I think that Parse should not load authentication adapters that have no configuration in the Parse config at all. Only those adapters should be loaded that the Parse admin explicitly specifies in the config. But of course this would be a backwards incompatible change and probably shouldn't be part of this PR. It's just my opinion.